### DISCUSSION TITLE
RDB + Autogroup testing

### DISCUSSION THREAD

**Discussion by hamadmarri** [hamadmarri/cacule-cpu-scheduler#41] [General]
Could you please test RDB + Autogroup

I attached 2 patch files

`cacule-rdb-autogroup.patch` is the full patch cacule+rdb+autogroup
`rdb-autogroup-testing.patch` is only rdb + autogroup <- in case you need to patch on top of your current cacule

[cacule-rdb-autogroup.zip](https://github.com/hamadmarri/cacule-cpu-scheduler/files/6844403/cacule-rdb-autogroup.zip)
[rdb-autogroup-testing.zip](https://github.com/hamadmarri/cacule-cpu-scheduler/files/6844404/rdb-autogroup-testing.zip)

Thank you

**Comment by JohnyPeaN** [comment]
@hamadmarri
I have tested it remotely on my home computer and have first observation.
Under load (compiling kernel) rdb doesn't use cores up to 100% and lets them partially run on idle, it can be seen in htop and other utils. Frequency scaling seems good, but i still use ondemand here. Might be some accounting problem, because user+sys are lower for the same work. But the job still took longer so the efficiency seems worse:

5.13.3-181-tkg-cacule RDB AUTOGROUP (output of dstat)
--total-cpu-usage-- -dsk/total- -net/total- ------memory-usage----- ---paging-- ---system-- ---procs--- 
usr sys idl wai stl| read  writ| recv  send| used  free  buff  cach|  in   out | int   csw |run blk new|
 76   7  16   0   0|   0     0 | 506B  820B|12.4G 48.5G  174M 10.6G|   0     0 |  14k 8982 | 25   0 414|
 75   8  17   0   0|   0     0 | 374B 2094B|12.2G 48.7G  174M 10.6G|   0     0 |  14k 9651 | 25   0 488|
 78   8  14   0   0|   0     0 | 126B  702B|12.1G 48.7G  174M 10.6G|   0     0 |  13k 8243 | 24   0 388|
 76   8  17   0   0|   0     0 | 186B  702B|12.0G 48.9G  174M 10.6G|   0     0 |  14k 9504 | 24   0 499|
 72   6  22   0   0|   0     0 |  66B  678B|12.8G 48.1G  174M 10.6G|   0     0 |  12k 8184 | 24   0 347|
 63   6  31   0   0|   0     0 | 309B  702B|12.3G 48.6G  174M 10.6G|   0     0 |  11k 7219 | 26   0 308|
 77   7  16   0   0|   0     0 | 126B  678B|12.4G 48.5G  174M 10.7G|   0     0 |  13k 8344 | 25   0 366|
 73   7  20   0   0|   0     0 |  66B  678B|12.4G 48.5G  174M 10.7G|   0     0 |  12k 7405 | 24   0 333|
real    11m50,965s
user    195m20,447s
sys     18m54,226s

5.13.3-181-tkg-cacule without RDB
 --total-cpu-usage-- -dsk/total- -net/total- ------memory-usage----- ---paging-- ---system-- ---procs---
usr sys idl wai stl| read  writ| recv  send| used  free  buff  cach|  in   out | int   csw |run blk new|
 91   8   1   0   0|   0     0 |1775B 2259B|10.9G 48.4G  153M 10.9G|   0     0 |  15k 7119 | 24   0 369|
 91   8   1   0   0|   0     0 |1649B 3375B|11.1G 48.2G  153M 10.9G|   0     0 |  14k 7104 | 51   0 355|
 90   9   2   0   0|   0     0 |2171B 7069B|11.2G 48.0G  153M 10.9G|   0     0 |  15k 8410 | 25   0 485|
 91   8   1   0   0|   0    24k|3598B 4114B|11.1G 48.2G  153M 10.9G|   0     0 |  15k 8275 | 25   0 388|
 91   8   1   0   0|   0     0 |1529B 2149B|10.9G 48.3G  153M 11.0G|   0     0 |  15k 7816 | 24   0 398|
 91   8   1   0   0|   0     0 | 246B  678B|10.9G 48.3G  153M 11.0G|   0     0 |  14k 6958 | 24   0 384|
 91   8   1   0   0|   0     0 |1709B 2165B|10.9G 48.4G  153M 11.0G|   0     0 |  16k 9673 | 24   0 373|
 91   8   1   0   0|   0     0 |1709B 2141B|11.0G 48.3G  153M 11.0G|   0     0 |  15k 7410 | 24   0 351|
real    10m34,440s
user    215m24,220s
sys     20m54,322s

I will test responsiveness later.

**Comment by ptr1337** [comment]
Faced the same issue by me at compiling or also at stress-ng.

**Comment by hamadmarri** [comment]
@JohnyPeaN @ptr1337 

Thank you for testing. I am also having this issue, and I can't yet find the reason why the poor balance. It could be because of:
* head_IS is not updated correctly or propagated through parents group 
* rdb_task function is not properly traversing the right children of the group

I can't think of other reasons. Please let me know if you have any proposed idea.

Thank you so much

**Comment by raykzhao** [comment]
Hi @hamadmarri,

I have noticed that the `select_task_rq_fair` has been reverted back to the old ways. Just in case someone would like to test the rdb-autogroup with the latest CacULE, I have made the synced patch (apply on top of the latest CacULE).
[rdb-autogroup-testing_synced.zip](https://github.com/hamadmarri/cacule-cpu-scheduler/files/6875569/rdb-autogroup-testing_synced.zip)

**Reply by hamadmarri** [reply]
> Hi @hamadmarri,
> 
> I have noticed that the `select_task_rq_fair` has been reverted back to the old ways. Just in case someone would like to test the rdb-autogroup with the latest CacULE, I have made the synced patch (apply on top of the latest CacULE).
> [rdb-autogroup-testing_synced.zip](https://github.com/hamadmarri/cacule-cpu-scheduler/files/6875569/rdb-autogroup-testing_synced.zip)

Hi @raykzhao 

Yes, since the CFS stats and accounting are used, it is better to use normal cacule/cfs way for `select_task_rq`. Does it have any effects on autogroup patch?

Thank you for providing the patch.

**Reply by raykzhao** [reply]
Hi @hamadmarri,

On my laptop with this patch the system feels slightly smoother than the previous rdb+autogroup. However, I'm not sure if this patch fixes the issues pointed out by @JohnyPeaN  and @ptr1337.

**Comment by hamadmarri** [comment]
I got some good rdb+autogroup version. It is finally balancing well as plain rdb. The performance also very good and some times better than the old plain rdb. Please test.

Git tree:
https://github.com/hamadmarri/linux/tree/cacule-5.13-rdb-autogroup-testing

All in one (cacule + rdb + autogroup):
https://github.com/hamadmarri/cacule-cpu-scheduler/blob/master/patches/CacULE/experimental/cacule-5.13-rdb-autogroup.patch

Thank you
